evm_arithmetization/src/cpu/kernel/asm/mpt/linked_list/linked_list.asm

```diff
@@ -0,0 +1,982 @@
+    PUSH %%after
+    %jump(store_initial_accounts)
+%%after:
+%endmacro
+
+/// Iterates over the inital account linked list and shallow copies
+/// the accounts, storing a pointer to the copied account in the node.
+global store_initial_accounts:
+    // stack: retdest
+    PUSH  @SEGMENT_ACCOUNTS_LINKED_LIST
+    %next_account
+loop_store_initial_accounts:
+    // stack: current_node_ptr
+    %get_trie_data_size
+    DUP2
+    MLOAD_GENERAL
+    // stack: current_addr, cpy_ptr, current_node_ptr, retdest
+    %eq_const(@U256_MAX)
+    %jumpi(store_initial_accounts_end)
+    DUP2
+    %increment
+    MLOAD_GENERAL
+    // stack: nonce_ptr, cpy_ptr, current_node_ptr, retdest
+    DUP1
+    %mload_trie_data // nonce
+    %append_to_trie_data
+    %increment
+    // stack: balance_ptr, cpy_ptr, current_node_ptr, retdest
+    DUP1
+    %mload_trie_data // balance
+    %append_to_trie_data
+    %increment // The storage_root_ptr is not really necessary
+    // stack: storage_root_ptr_ptr, cpy_ptr, current_node_ptr, retdest
+    DUP1
+    %mload_trie_data // storage_root_ptr
+    %append_to_trie_data
+    %increment
+    // stack: code_hash_ptr, cpy_ptr, current_node_ptr, retdest
+    %mload_trie_data // code_hash
+    %append_to_trie_data
+    // stack: cpy_ptr, current_node_ptr, retdest
+    DUP2
+    %add_const(2)
+    SWAP1
+    MSTORE_GENERAL // Store cpy_ptr
+    %next_account
+    %jump(loop_store_initial_accounts)
+
+store_initial_accounts_end:
+    %pop2
+    JUMP
+
+
+%macro insert_account_to_linked_list
+    %stack (addr, ptr) -> (addr, ptr, %%after)
+    %jump(insert_account_to_linked_list)
+%%after:
+    // stack: payload_ptr
+%endmacro
+
+%macro insert_account_with_overwrite
+    %stack (addr, ptr) -> (addr, ptr, %%after)
+    %jump(insert_account_with_overwrite)
+%%after:
+    // stack: payload_ptr
+%endmacro
+
+%macro insert_account_to_linked_list_no_return
+    %insert_account_to_linked_list
+    POP
+%endmacro
+
+%macro insert_account_with_overwrite_no_return
+    %insert_account_with_overwrite
+    POP
+%endmacro
+
+// Multiplies the value at the top of the stack, denoted by ptr/4, by 4
+// and aborts if ptr/4 >= mem[@GLOBAL_METADATA_ACCOUNTS_LINKED_LIST_LEN]/4
+// This way, 4*ptr/4 must be pointing to the beginning of a node.
+// TODO: Maybe we should check here if the node has been deleted.
+%macro get_valid_account_ptr
+    // stack: ptr/4
+    DUP1
+    PUSH 4
+    %mload_global_metadata(@GLOBAL_METADATA_ACCOUNTS_LINKED_LIST_LEN)
+    // By construction, both @SEGMENT_ACCESSED_STORAGE_KEYS and the unscaled list len
+    // must be multiples of 4
+    DIV
+    // stack: @SEGMENT_ACCESSED_STORAGE_KEYS/4 + accessed_strg_keys_len/4, ptr/4, ptr/4
+    %assert_gt
+    %mul_const(4)
+%endmacro
+
+/// Inserts the account addr and payload pointer into the linked list if it is not already present.
+/// Returns `payload_ptr` if the account was inserted, `original_ptr` if it was already present.
+global insert_account_to_linked_list:
```

Comment:
From what I'm seeing, we make use of  `payload_ptr` in only one context: to return `0` if the account is not found. We might as well hardcode `0` as the value if the account wasn't already present.

Reply:
I removed `insert_account_to_linked_list`, along with some macros, and made other insertion methods not return the payload pointer.